### DISCUSSION TITLE
Custom Traits Not Properly Working for MJML

### DISCUSSION THREAD

**Discussion by RakulAgn** [GrapesJS/grapesjs#4795] [Q&A]
```
setTimeout(() => {
    editor.DomComponents.addType("mj-image", {
      isComponent: (el: any) => el.tagName === "MJ-IMAGE",
      model: {
        defaults: {
          traits: [
            {
              type: "mjchange",
              label: " ",
              name: "mjchange",
            },
            {
              type: "href",
              label: "Link",
              name: "href",
            },
            {
              type: "alt",
              label: "Alt",
              name: "alt",
            },
          ],
        },
      },
    });

  editor.TraitManager.addType("mjchange", {
    noLabel: true,
    createInput({}) {
      let selectedSrc = editor.getSelected();

      let src = selectedSrc!.attributes.attributes!.src;

      const toggleModal = () => {
        editor.runCommand("open-assets", {
          target: editor.getSelected(),
        });
      };
      const el = document.createElement("div");
      el.setAttribute("class", "image-trait-preview");
      el.innerHTML = `<img src="${src}" style="width: 100%; height:auto;background:#f9f9f9;" id="gjs_img_preview_logo_rtl"/>
                <button type="submit" class="btn btn-primary btn-block btn-sm" id="chg-img-trait-btn">Change</button>`;

      const inputType = el.querySelector("#chg-img-trait-btn");
      const imgBox = el.querySelector("#gjs_img_preview_logo_rtl");

      imgBox!.addEventListener("click", toggleModal);
      inputType!.addEventListener("click", toggleModal);

      return el;
    },
  });

  editor.on("modal:open", (component) => {
    const $ = editor.$;
    editor.AssetManager.open({
      types: ["mj-image"],
      select(assets, complete) {
        const selected = editor.getSelected();    
        if (selected && selected.is("mj-image")) {
          selected.addAttributes({ src: assets.getSrc() });
          $("#gjs_img_preview_logo_rtl").attr("src", assets.getSrc());
          complete && editor.AssetManager.close();
        }
      },
    });
```
This Method Works Fine for Normal HTML. But When I Tried With MJML For MJ-IMAGE Tag
The Src is Updated in TR Tag but not updated in img tag inside the TR tag
The Below Image 
![Screenshot from 2022-12-21 13-32-13(1)](https://user-images.githubusercontent.com/119593989/208852368-33fc6704-88e4-4ac0-8f9c-9c022efd67cb.jpg)

**Comment by artf** [comment]
try to update `selected.addAttributes({ src: assets.getSrc() });` with `selected.set({ src: assets.getSrc() })`

**Comment by zeeshan-algo** [comment]
@RakulAgn Did you solve this issue? If yes than tell me how u solved this?

**Reply by RakulAgn** [reply]
try to update selected.addAttributes({ src: assets.getSrc() }); with selected.set({ src: assets.getSrc() })

**Comment by zeeshan-algo** [comment]
@RakulAgn can you plzz share the whole code regarding picture

**Reply by RakulAgn** [reply]
![image](https://user-images.githubusercontent.com/119593989/223041832-38261b83-8113-42e2-b48e-54298f2046e3.png)

You Can Remove the Condition.. its required for my Project

**Reply by RakulAgn** [reply]
You have to define Custom DomComponents. For that DomComponents You have to create Custom Traits and its functionality .. You may see that `mj-image` and `mj-social-elements` are default gjs types in grapejs so you can use them to create the functionality of what you expecting from picture traits

**Comment by zeeshan-algo** [comment]
@RakulAgn i am getting the same error as you described above . when i upload the image from trait is show the src of the image  trait update  but the src is not updated at canvas . i tried using this code `selected.set({ src: assets.getSrc() })`  but it does not work it show a error that one argument val is missing
![screen2](https://user-images.githubusercontent.com/115087247/223057322-80e6093f-378d-48f6-9730-8ee7d056ee65.png)
![screen3](https://user-images.githubusercontent.com/115087247/223057334-bd550844-4f27-40c9-abc4-191582e59466.png)
![screen 1](https://user-images.githubusercontent.com/115087247/223057337-e8eec090-8bbc-46ad-98e0-f6d23810c5b4.png)

**Reply by RakulAgn** [reply]
Can you share you code

**Reply by zeeshan-algo** [reply]
@RakulAgn same code as yours

**Comment by zeeshan-algo** [comment]
@RakulAgn   editor.DomComponents.addType("mj-image", {
        isComponent: (el: any) => el.tagName === "MJ-IMAGE",
        model: {
          defaults: {
            traits: [
              {
                type: "mjchange",
                label: " ",
                name: "mjchange",
              },
              {
                type: "href",
                label: "Link",
                name: "href",
              },
              {
                type: "alt",
                label: "Alt",
                name: "alt",
              },
            ],
          },
        },
      });



  editor.TraitManager.addType("mjchange", {
        noLabel: true,
        createInput({}) {
          let selectedSrc = editor.getSelected();
    
          let src = selectedSrc!.attributes.attributes!.src;
          const toggleModal = () => {
            editor.runCommand("open-assets", {
              target: editor.getSelected(),
            });
          };
          const el = document.createElement("div");
          el.setAttribute("class", "image-trait-preview");
          el.innerHTML = `<img src="${src}" style="width: 100%; height:auto;background:#f9f9f9;" id="gjs_img_preview_logo_rtl"/>
                    <button type="submit" id="chg-img-trait-btn">Upload</button>`;
    
          const inputType = el.querySelector("#chg-img-trait-btn");
          const imgBox = el.querySelector("#gjs_img_preview_logo_rtl");
    
          imgBox!.addEventListener("click", toggleModal);
          inputType!.addEventListener("click", toggleModal);
    
          return el;
        },
      });
      
      
      
      
       editor.on("modal:open", (component) => {
        const $ = editor.$;
        editor.AssetManager.open({
          types: ["mj-image"],
          select(assets, complete) {
            const selected = editor.getSelected();    
            if (selected && selected.is("mj-image")) {
              console.log("selected----->", selected);
              console.log("assets----->", assets.getSrc());
             



              selected.set("src",{ src: assets.getSrc() });
              // selected.addAttributes({ src: assets.getSrc() });
              $("#gjs_img_preview_logo_rtl").attr("src", assets.getSrc());
              complete && editor.AssetManager.close();
            }
          },
        });
      });

**Reply by zeeshan-algo** [reply]
basically i am unable to set the src with assets.getSrc().

**Reply by RakulAgn** [reply]
`selected.set("src",{ src: assets.getSrc() });` replace `selected.set("src", assets.getSrc());`

**Reply by zeeshan-algo** [reply]
nope still image not shown in canvas and src is not updated

**Reply by RakulAgn** [reply]
dude have you installed jquery ? properly

**Reply by RakulAgn** [reply]
is that any error is showing ?
